### ISSUE TITLE
TypeError: Cannot read property '$new' of undefined

### ISSUE THREAD

**Issue by bfitzpat** [asciidoctor/opal-node-runtime#58]
We are using the asciidoctor npm library and suddenly have started to see issues loading opal in our vscode extension code. I saw the same issue happen with the https://marketplace.visualstudio.com/items?itemName=asciidoctor.asciidoctor-vscode extension as well. 

The stack trace appears when we attempt to instantiate a reference to asciidoctor (i.e. `const asciidoctor = require('asciidoctor')();`). Opal is obviously a transitive dependency at that point, so I'm not sure what would be going on there. 

Any ideas what might be causing this?

```
[2021-02-24 14:24:34.832] [exthost] [error] TypeError: Cannot read property '$new' of undefined
	at Object.bridge (/home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:1154:32)
	at /home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:527:12
	at /home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:4919:16
	at /home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:5072:5
	at Object.load (/home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:2311:7)
	at constructor.<anonymous> (/home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:2339:17)
	at /home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:5439:8
	at Object.load (/home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:2311:7)
	at constructor.<anonymous> (/home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:2339:17)
	at /home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:19939:8
	at Object.<anonymous> (/home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/opal.js:19949:3)
	at Module._compile (internal/modules/cjs/loader.js:1199:30)
	at Object..js (internal/modules/cjs/loader.js:1220:10)
	at Module.load (internal/modules/cjs/loader.js:1039:32)
	at internal/modules/cjs/loader.js:932:14
	at Function.<anonymous> (electron/js2c/asar_bundle.js:5:12738)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:27152)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:23904)
	at Function._load (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:94:62159)
	at Module.require (internal/modules/cjs/loader.js:1079:19)
	at require (internal/modules/cjs/helpers.js:72:18)
	at Object.<anonymous> (/home/bfitzpat/github/vscode-didact/node_modules/asciidoctor-opal-runtime/src/index.js:2:1)
	at Module._compile (internal/modules/cjs/loader.js:1199:30)
	at Object..js (internal/modules/cjs/loader.js:1220:10)
	at Module.load (internal/modules/cjs/loader.js:1039:32)
	at internal/modules/cjs/loader.js:932:14
	at Function.<anonymous> (electron/js2c/asar_bundle.js:5:12738)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:27152)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:23904)
	at Function._load (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:94:62159)
	at Module.require (internal/modules/cjs/loader.js:1079:19)
	at require (internal/modules/cjs/helpers.js:72:18)
	at Object.<anonymous> (/home/bfitzpat/github/vscode-didact/node_modules/@asciidoctor/core/dist/node/asciidoctor.js:2:14)
	at Module._compile (internal/modules/cjs/loader.js:1199:30)
	at Object..js (internal/modules/cjs/loader.js:1220:10)
	at Module.load (internal/modules/cjs/loader.js:1039:32)
	at internal/modules/cjs/loader.js:932:14
	at Function.<anonymous> (electron/js2c/asar_bundle.js:5:12738)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:27152)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:23904)
	at Function._load (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:94:62159)
	at Module.require (internal/modules/cjs/loader.js:1079:19)
	at require (internal/modules/cjs/helpers.js:72:18)
	at Object.<anonymous> (/home/bfitzpat/github/vscode-didact/out/asciidocUtils.js:20:21)
	at Module._compile (internal/modules/cjs/loader.js:1199:30)
	at Object..js (internal/modules/cjs/loader.js:1220:10)
	at Module.load (internal/modules/cjs/loader.js:1039:32)
	at internal/modules/cjs/loader.js:932:14
	at Function.<anonymous> (electron/js2c/asar_bundle.js:5:12738)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:27152)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:23904)
	at Function._load (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:94:62159)
	at Module.require (internal/modules/cjs/loader.js:1079:19)
	at require (internal/modules/cjs/helpers.js:72:18)
	at Object.<anonymous> (/home/bfitzpat/github/vscode-didact/out/extensionFunctions.js:37:25)
	at Module._compile (internal/modules/cjs/loader.js:1199:30)
	at Object..js (internal/modules/cjs/loader.js:1220:10)
	at Module.load (internal/modules/cjs/loader.js:1039:32)
	at internal/modules/cjs/loader.js:932:14
	at Function.<anonymous> (electron/js2c/asar_bundle.js:5:12738)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:27152)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:23904)
	at Function._load (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:94:62159)
	at Module.require (internal/modules/cjs/loader.js:1079:19)
	at require (internal/modules/cjs/helpers.js:72:18)
	at Object.<anonymous> (/home/bfitzpat/github/vscode-didact/out/extension.js:30:28)
	at Module._compile (internal/modules/cjs/loader.js:1199:30)
	at Object..js (internal/modules/cjs/loader.js:1220:10)
	at Module.load (internal/modules/cjs/loader.js:1039:32)
	at internal/modules/cjs/loader.js:932:14
	at Function.<anonymous> (electron/js2c/asar_bundle.js:5:12738)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:27152)
	at Function.<anonymous> (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:23904)
	at Function._load (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:94:62159)
	at Module.require (internal/modules/cjs/loader.js:1079:19)
	at require (internal/modules/cjs/helpers.js:72:18)
	at Function.t [as __$__nodeRequire] (/usr/share/code/resources/app/out/vs/loader.js:4:101)
	at u._loadCommonJSModule (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:104:25310)
	at u._doActivateExtension (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:90:11562)
	at u._activateExtension (/usr/share/code/resources/app/out/vs/workbench/services/extensions/node/extensionHostProcess.js:90:10621)
	at processTicksAndRejections (internal/process/task_queues.js:97:5)
	at async Promise.all (index 2)
	at async Promise.all (index 0)
```

**Comment by bfitzpat** [comment]
Our extension is https://github.com/redhat-developer/vscode-didact (not the other asciidoc extension) and our current dependency list in package.json looks like this. Perhaps it's a version clash somewhere?

```
	"devDependencies": {
		"@types/chai": "^4.2.15",
		"@types/glob": "^7.1.3",
		"@types/mocha": "^8.2.1",
		"@types/node": "^14.14.31",
		"@types/sinon": "^9.0.10",
		"@types/vscode": "^1.53.0",
		"@typescript-eslint/eslint-plugin": "^4.15.1",
		"@typescript-eslint/parser": "^4.15.1",
		"chai": "^4.3.0",
		"eslint": "^7.20.0",
		"mocha": "^8.3.0",
		"mocha-jenkins-reporter": "^0.4.4",
		"sinon": "9.2.4",
		"typescript": "^4.1.5",
		"vscode-test": "^1.5.1"
	},
	"dependencies": {
		"@types/download": "^6.2.4",
		"@types/markdown-it": "^12.0.1",
		"@types/request-promise": "^4.1.47",
		"asciidoctor": "^2.2.1",
		"async-wait-until": "2.0.3",
		"base-64": "^1.0.0",
		"download": "^8.0.0",
		"glob": "^7.1.6",
		"markdown-it": "^12.0.4",
		"markdown-it-attrs": "^4.0.0",
		"markdown-it-task-lists": "^2.1.1",
		"node-fetch": "^2.6.1",
		"node-html-parser": "^2.1.0",
		"tmp": "^0.2.1",
		"url-parse": "^1.5.1"
	}
```

**Comment by ggrossetie** [comment]
Do you have both extensions installed? Opal can only be instantiated once maybe that's why you are seeing this error. Could you please try to start VS code with only your extension installed?

**Comment by ggrossetie** [comment]
Please note that you should use `@asciidoctor/core` instead of `asciidoctor` which includes `@asciidoctor/core` but also `@asciidoctor/cli` (which I believe is not necessary in your context). If you do that you will need to update the `require` statement.

**Comment by bfitzpat** [comment]
> Do you have both extensions installed? Opal can only be instantiated once maybe that's why you are seeing this error. Could you please try to start VS code with only your extension installed?

This is likely the issue. We're both loading opal. In the VS Code extension world I'm not sure how to get around that. Have to do some research here. Thank you for the insight!

**Comment by bfitzpat** [comment]
@Mogztter that is exactly the issue. We are trying to coexist with https://marketplace.visualstudio.com/items?itemName=asciidoctor.asciidoctor-vscode and use the asciidoctor parser. 

Is there any way for me to instantiate an asciidoctor instance without clashing on the `opal` issue?

as soon as we issue a `const asciidoctor = require('@asciidoctor/core')();` we are dead in the water.

**Comment by bfitzpat** [comment]
As a bit of background... Ultimately we would like to provide additional functionality to what you get with the asciidoctor extension https://marketplace.visualstudio.com/items?itemName=asciidoctor.asciidoctor-vscode - but we are currently unable to do that due to this limitation. We add some new functionality via our vscode-didact extension (additional code completion for links that leverage the underlying vscode command API to accomplish a type of interactive tutorial process within vscode). Our extension also offers a custom preview (which is why we need adoc -> html rendering as provided by the adoc library). 

It would be great if we could just rely on the asciidoctor.asciidoctor-vscode extension as a dependency and peacefully coexist but that doesn't seem possible at the moment.

**Comment by ggrossetie** [comment]
@bfitzpat Thanks for your feedback!
I'm not entirely sure why this exception is thrown because the following code run just fine in Node.js:

```js
var asciidoctor = require('asciidoctor')()
  
var asciidoctor = require('@asciidoctor/core')()

console.log(asciidoctor.convert('Hello *world*'))
```

You can also `require('asciidoctor-opal-runtime')` multiple times...

As far as I know, Electron compiles the JavaScript code (or at least creates a bundle) and uses a custom CommonJS loader... I'm not sure what exactly is happening but you should try to add some debug statements/break-points when your extension loads correctly and when it does not, maybe you will spot key differences.

**Comment by ggrossetie** [comment]
Do you see this warning `Opal already loaded. Loading twice can cause troubles, please fix your setup.` somewhere in the debug console?

**Comment by ggrossetie** [comment]
@bfitzpat If you can provide a _simple_ reproduction case, I can help you troubleshoot this issue.

**Comment by ggrossetie** [comment]
Good news, I can reproduce this issue and I have a potential fix :tada:

**Comment by bfitzpat** [comment]
Sorry @Mogztter was on PTO for a couple of days. Looking forward to seeing your potential fix!

**Comment by ggrossetie** [comment]
fixed in 61b576c
